Made BGG API calls more efficient (#14)

ironsigma opened the pull request:
:wrench: Updated BBGPayloadSearchResults to accepts indexing.
:hammer: Added two new methods to BBGService interface to support new fetch method.
:wrench: Updated DefaultBGGService with new batch API implementation
:wrench: Updated GameDAO to accept indexing to copy the specified game record
:hammer: Added query to GameRepo to fetch latest game ID from the database
:wrench: Updated ScheduledTasks to call new API

:warning: You will need to wipe your database, since calls are sequential now, it checks the DB for the last saved record and picks up from there.

ironsigma committed:
Made BGG API calls more efficient (#14)

ironsigma commented on src/main/java/org/bunge/bgginfo/bggintegration/model/BGGPayloadSearchResults.java:
```diff
@@ -28,49 +27,59 @@ public void setTermsofuse(String termsofuse) {
-                    rec = Integer.parseInt(rslt.numVotes);
-                else if (rslt.value.equals("Not Recommended"))
-                    nRec = Integer.parseInt(rslt.numVotes);
+            if (rs.result != null) {
```
This was throwing a NPE, added a check.

mattb555 commented on src/main/java/org/bunge/bgginfo/scheduled/ScheduledTasks.java:
```diff
@@ -1,27 +1,19 @@
-	private static final Logger log = LoggerFactory.getLogger(ScheduledTasks.class);
- 
-    @Scheduled(fixedRate = 3000)
+    @Scheduled(fixedRate = 30000)
```
As far as I can tell there's no need to slow the rate down like this. I haven't had any calls fails at 3000ms.

mattb555 replied:
Please change it back for approval.

mattb555 commented on src/main/java/org/bunge/bgginfo/bggintegration/service/DefaultBGGService.java:
```diff
@@ -1,33 +1,61 @@
+    private static final String BGG_URL_TEMPLATE = "https://www.boardgamegeek.com/xmlapi2";
+    private static final int BGG_BATCH_SIZE = 20;
+    private static final int BGG_MAX_GAME_ID = 5000000;
 
```
So something that I might not have explained well is the point of the BGG Service is to run perpetually, so that the any updates that need to be made to the data can be done. The end goal once I'm happy that it's fully featured is to host it on something like Heroku so it will need to do that.

I'll leave it up to you if you want make that change or not because as it stands this PR is a significant improvement over the original project in multiple aspects so I have no qualms approving it as is.

mattb555 reviewed (changes requested):
I left 2 comments. If you want to make the bigger change I mentioned I would appreciate it, otherwise please fix the smaller one and then I can get this merged.

mattb555 commented:
@all-contributors please add @ironsigma for code

ironsigma committed:
Changing back scheduled timer (#14)

ironsigma commented:
Fixed the timer, I changed it for testing and didn't think to put it back.

How do you see the perpetual run? Just reset the counter to 1 once we hit max, and update records if necessary?

mattb555 commented:
Reset at one works, record should probably be updated regardless, it will be easier than checking if all the fields are the same.

ironsigma committed:
Making API fetch restart when it reaches max game ID (#14)

ironsigma commented:
Updated fetch API to restart when it reaches max ID

ironsigma commented:
@mattb555 I've updated the code to restart after reaching max ID, let me know if there's anything else you want me to add.
